optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,350 @@
+        else:
+            if self._qmc_type == "sobol":
+                qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            elif self._qmc_type == "halton":
+                qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "sobol" and "halton".'
+                )
+                raise ValueError(message)
+
+        assert isinstance(qmc_engine, scipy.stats.qmc.QMCEngine)
+
+        forward_size = sample_id - qmc_engine.num_generated  # `sample_id` starts from 0.
+        qmc_engine.fast_forward(forward_size)
+        sample = qmc_engine.random(1)
+        self._cached_qmc_engine = qmc_engine
+
+        return sample
+
+    def _find_sample_id(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> int:
+
+        qmc_id = ""
+        qmc_id += self._qmc_type
+        qmc_id += str(search_space)
+        # Sobol/Halton sequences without scrambling do not use seed.
+        if not self._scramble:
+            qmc_id += str(self._seed)
+        hashed_qmc_id = hash(qmc_id)
+        key_qmc_id = f"qmc ({hashed_qmc_id})'s last sample id"
+
+        # TODO(kstoneriv3): Following try-except block assumes that the block is
+        # an atomic transaction. This ensures that each sample_id is sampled at least once.
+        try:
+            sample_id = study._storage.get_study_system_attrs(study._study_id)[key_qmc_id]
+            sample_id += 1
+            study._storage.set_study_system_attr(study._study_id, key_qmc_id, sample_id)
+        except KeyError:
+            study._storage.set_study_system_attr(study._study_id, key_qmc_id, 0)
+            sample_id = 0
+
+        return sample_id
+
+    def _is_engine_cached(self, d: int, sample_id: int) -> bool:
+
+        if not isinstance(self._cached_qmc_engine, scipy.stats.qmc.QMCEngine):
+            return False
+        else:
+            is_cached = True
+            is_cached &= self._cached_qmc_engine.rng_seed == self._seed
+            is_cached &= self._cached_qmc_engine.d == d
+            is_cached &= self._cached_qmc_engine.num_generated <= sample_id
+
+            # TODO(kstoneriv3): Maybe we should assume that
+            # `_qmc_type` does not change for simplicity.
+            if self._qmc_type == "sobol":
+                is_cached &= self._cached_qmc_engine.__class__.__name__ == "Sobol"
+            elif self._qmc_type == "halton":
+                is_cached &= self._cached_qmc_engine.__class__.__name__ == "Halton"
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "sobol" and "halton".'
+                )
+                raise ValueError(message)
```

Comment:
Do you believe I can omit this section? Under the assumption that the user does not update the `self._qmc_type`, we can get rid of this section.